grpc/src/main/java/com/linecorp/armeria/internal/server/grpc/AbstractServerCall.java

```diff
@@ -209,66 +208,78 @@ protected final void maybeCancel() {
         final Metadata metadata = generateMetadataFromThrowable(exception);
-        close(GrpcStatus.fromThrowable(exceptionHandler, ctx, exception, metadata), metadata, exception);
+        final Status status = GrpcStatus.fromThrowable(exceptionHandler, ctx, exception, metadata);
+        close(new ServerStatusAndMetadata(status, metadata, cancelled), exception);
     }
 
     @Override
     public final void close(Status status, Metadata metadata) {
-        close(GrpcStatus.fromExceptionHandler(exceptionHandler, ctx, status, metadata), metadata, null);
+        final ServerStatusAndMetadata statusAndMetadata =
+                new ServerStatusAndMetadata(GrpcStatus.fromExceptionHandler(exceptionHandler, ctx,
+                                                                            status, metadata),
+                                            metadata, false);
+        close(statusAndMetadata, null);
     }
 
-    private void close(Status status, Metadata metadata, @Nullable Throwable exception) {
+    private void close(ServerStatusAndMetadata statusAndMetadata, @Nullable Throwable exception) {
         if (ctx.eventLoop().inEventLoop()) {
-            doClose(status, metadata, exception);
+            doClose(statusAndMetadata, exception);
         } else {
             ctx.eventLoop().execute(() -> {
-                doClose(status, metadata, exception);
+                doClose(statusAndMetadata, exception);
             });
         }
     }
 
-    private void doClose(Status status, Metadata metadata, @Nullable Throwable exception) {
+    private void doClose(ServerStatusAndMetadata statusAndMetadata, @Nullable Throwable exception) {
         maybeLogFailedRequestContent(exception);
+        Status status = statusAndMetadata.status();
+        final Metadata metadata = statusAndMetadata.metadata();
         if (isCancelled()) {
             // No need to write anything to client if cancelled already.
-            closeListener(status, metadata, false, true);
+            statusAndMetadata.completed(false);
+            statusAndMetadata.setResponseContent(true);
+            closeListener(statusAndMetadata);
             return;
         }
 
         if (status.getCode() == Code.CANCELLED && status.getCause() instanceof ClosedStreamException) {
-            closeListener(status, metadata, false, true);
+            statusAndMetadata.completed(false);
+            statusAndMetadata.setResponseContent(true);
+            closeListener(statusAndMetadata);
             return;
         }
 
         checkState(!closeCalled, "call already closed. status: %s, exception: %s",
                    status, exception);
         closeCalled = true;
 
-        boolean completed = true;
         if (status.getCode() == Code.CANCELLED && status.getCause() instanceof RequestTimeoutException) {
             // A call was finished by a timeout scheduler, not a user.
-            completed = false;
+            statusAndMetadata.completed(false);
```

Comment:
Question: Could this logic be moved inside `ServerStatusAndMetadata`?

Reply:
I'm not sure if this is the same comment as https://github.com/line/armeria/pull/5273#discussion_r1387629083.

I'd rather not have a dependency between `completed` and `cancelled` since they mean two different things. Instead, I just removed the condition altogether and set `cancelled=true` when server-side timeout is exceeded (thanks to the `ServerStatusAndMetadata` suggestion 👍 )